### PR TITLE
Add JVMTI synchronization in JVM_VirtualThreadHideFrames

### PR DESCRIPTION
**Pull request by babsingh**
`J9VMThread->threadObject` can be modified between
`JVM_VirtualThreadHideFrames`'s first invocation with `hide=true` and
second invocation with `hide=false`. Synchronization to prevent JVMTI
operations is acquired between these two invocations. This prevents
JVMTI functions to see an unstable `J9VMThread->threadObject`.

Related: eclipse-openj9#17865
Related: eclipse-openj9#17869
Related: eclipse-openj9#18370

### CODE REVIEW

**Comment by babsingh** [comment]
@fengxue-IS @tajila Requesting your review.

This is the last PR to address https://github.com/eclipse-openj9/openj9/issues/17865, https://github.com/eclipse-openj9/openj9/issues/17869 and https://github.com/eclipse-openj9/openj9/issues/18370.

File: runtime/j9vm/javanextvmi.cpp [hunk 1]
```diff
@@ -577,13 +577,28 @@ JNIEXPORT void JNICALL
...
-		Assert_SC_true(J9_ARE_NO_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
+		Assert_SC_true(!hiddenFrames && (vThreadObj == J9_JNI_UNWRAP_REFERENCE(vthread)));
+		enterVThreadTransitionCritical(currentThread, vthread);
 	} else {
-		Assert_SC_true(J9_ARE_ALL_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
+		Assert_SC_true(hiddenFrames);
+		exitVThreadTransitionCritical(currentThread, vThreadObj);
 	}
+
 	VM_VMHelpers::virtualThreadHideFrames(currentThread, hide);
```
Comment (fengxue-IS) [inline runtime/j9vm/javanextvmi.cpp:600]: flag toggle should be moved inside the if check to maintain correct order and avoid sync issues in the future
ie:
enterVThreadTransitionCritical
virtualThreadHideFrames(true)
virtualThreadHideFrames(false)
exitVThreadTransitionCritical

**Review by fengxue-IS — CHANGES_REQUESTED** [review]

File: runtime/j9vm/javanextvmi.cpp (hunk 1 above)
Comment (babsingh) [reply runtime/j9vm/javanextvmi.cpp]: Fixed `JVM_VirtualThreadHideFrames`. Also, addressed a similar issue in `virtualThreadMountBegin`.

File: runtime/j9vm/javanextvmi.cpp [hunk 2]
```diff
@@ -351,6 +349,8 @@ virtualThreadMountBegin(JNIEnv *env, jobject thread)
 		enterVThreadTransitionCritical(currentThread, thread);
 		threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
 	}
+
+	VM_VMHelpers::virtualThreadHideFrames(currentThread, JNI_TRUE);
```
Comment (babsingh) [inline runtime/j9vm/javanextvmi.cpp:353]: Moved it after the suspend loop; we should wait for the virtual thread to be resumed before invoking this function.

File: runtime/j9vm/javanextvmi.cpp [hunk 3]
```diff
@@ -377,6 +375,8 @@ virtualThreadMountEnd(JNIEnv *env, jobject thread)
 				J9VMJDKINTERNALVMCONTINUATION_VMREF(currentThread, continuationObj));
 	}
 
+	VM_VMHelpers::virtualThreadHideFrames(currentThread, JNI_FALSE);
```
Comment (babsingh) [inline runtime/j9vm/javanextvmi.cpp:378]: Moved it closer `exitVThreadTransitionCritical`, which is closer to the end phase.

File: runtime/j9vm/javanextvmi.cpp [hunk 4]
```diff
@@ -577,13 +578,31 @@ JNIEXPORT void JNICALL
...
-		Assert_SC_true(J9_ARE_NO_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
-	} else {
-		Assert_SC_true(J9_ARE_ALL_BITS_SET(currentThread->privateFlags, J9_PRIVATE_FLAGS_VIRTUAL_THREAD_HIDDEN_FRAMES));
+		Assert_SC_true(!hiddenFrames && (vThreadObj == J9_JNI_UNWRAP_REFERENCE(vthread)));
+		enterVThreadTransitionCritical(currentThread, vthread);
 	}
+
 	VM_VMHelpers::virtualThreadHideFrames(currentThread, hide);
+
+	if (!hide) {
```
Comment (babsingh) [inline runtime/j9vm/javanextvmi.cpp:600]: Split into two `if` statements so that `virtualThreadHideFrames` is only invoked once.

**Review by fengxue-IS — APPROVED** [review]
LGTM

**Comment by babsingh** [comment]
jenkins test sanity,extended.openjdk zlinux jdk21

**Comment by babsingh** [comment]
> https://openj9-jenkins.osuosl.org/job/Test_openjdk21_j9_extended.openjdk_s390x_linux_Personal/7/consoleFull

Only known failures are seen in the builds, which are unrelated to this PR:
- https://github.com/eclipse-openj9/openj9/issues/18426
- https://github.com/eclipse-openj9/openj9/issues/14948

**Comment by babsingh** [comment]
@tajila PR builds look good. Requesting your review.

**Review by tajila — APPROVED** [review]

**Comment by tajila** [comment]
@babsingh You'll need to triple deliver to 0.42 and 0.43

**Comment by babsingh** [comment]
Backports: 
- #18494
- #18495

**Pull request merged**